File: apps/api/src/api/schemas/applications.py
"""
applications.py - Pydantic schemas for Applications Tracker endpoints.
"""

from enum import Enum
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class ApplicationStatus(str, Enum):
    saved = "saved"
    cv_ready = "cv_ready"
    applied = "applied"
    follow_up = "follow_up"
    interview = "interview"
    rejected = "rejected"
    won = "won"
    archived = "archived"


class ApplicationItem(BaseModel):
    id: str
    user_id: Optional[str] = None
    offer_id: str
    offer_title: Optional[str] = None
    offer_company: Optional[str] = None
    offer_city: Optional[str] = None
    offer_country: Optional[str] = None
    status: ApplicationStatus
    source: str = "manual"
    note: Optional[str] = None
    next_follow_up_date: Optional[str] = None
    current_cv_cache_key: Optional[str] = None
    current_letter_cache_key: Optional[str] = None
    created_at: str
    updated_at: str
    applied_at: Optional[str] = None
    last_status_change_at: Optional[str] = None
    # Preparatory field for AI strategy layer (v0 — not yet LLM-generated)
    strategy_hint: Optional[str] = None


class ApplicationCreate(BaseModel):
    offer_id: str
    status: ApplicationStatus
    source: str = "manual"
    note: Optional[str] = None
    next_follow_up_date: Optional[str] = None


class ApplicationUpdate(BaseModel):
    status: Optional[ApplicationStatus] = None
    note: Optional[str] = None
    next_follow_up_date: Optional[str] = None
    strategy_hint: Optional[str] = None


class ApplicationListResponse(BaseModel):
    items: List[ApplicationItem] = Field(default_factory=list)


class ApplicationHistoryItem(BaseModel):
    id: str
    application_id: str
    from_status: Optional[str] = None
    to_status: str
    changed_at: str
    note: Optional[str] = None


class ApplicationHistoryResponse(BaseModel):
    items: List[ApplicationHistoryItem] = Field(default_factory=list)


class PrepareRequest(BaseModel):
    profile: Optional[Dict[str, Any]] = None
    enrich_llm: int = Field(default=0, ge=0, le=1)


class PrepareResponse(BaseModel):
    ok: bool
    application_id: str
    offer_id: str
    run_id: str
    cv_cache_key: Optional[str] = None
    letter_cache_key: Optional[str] = None
    status: str
    warnings: List[str] = Field(default_factory=list)

File: apps/web/src/api/applications.ts
import { apiFetch } from "../lib/api";

const API_BASE =
  import.meta.env.VITE_API_BASE_URL || import.meta.env.VITE_API_URL || "";

export type ApplicationStatus =
  | "saved"
  | "cv_ready"
  | "applied"
  | "follow_up"
  | "interview"
  | "rejected"
  | "won"
  | "archived";

export interface ApplicationItem {
  id: string;
  user_id: string | null;
  offer_id: string;
  offer_title: string | null;
  offer_company: string | null;
  offer_city: string | null;
  offer_country: string | null;
  status: ApplicationStatus;
  source: string;
  note: string | null;
  next_follow_up_date: string | null;
  current_cv_cache_key: string | null;
  current_letter_cache_key: string | null;
  created_at: string;
  updated_at: string;
  applied_at: string | null;
  last_status_change_at: string | null;
  strategy_hint: string | null;
}

export interface ApplicationListResponse {
  items: ApplicationItem[];
}

export interface ApplicationHistoryItem {
  id: string;
  application_id: string;
  from_status: string | null;
  to_status: string;
  changed_at: string;
  note: string | null;
}

export interface ApplicationHistoryResponse {
  items: ApplicationHistoryItem[];
}

export interface ApplicationCreatePayload {
  offer_id: string;
  status: ApplicationStatus;
  source?: string;
  note?: string | null;
  next_follow_up_date?: string | null;
}

export interface ApplicationUpdatePayload {
  status?: ApplicationStatus;
  note?: string | null;
  next_follow_up_date?: string | null;
  strategy_hint?: string | null;
}

export interface PreparePayload {
  profile?: Record<string, unknown>;
  enrich_llm?: 0 | 1;
}

export interface PrepareResponse {
  ok: boolean;
  application_id: string;
  offer_id: string;
  run_id: string;
  cv_cache_key: string | null;
  letter_cache_key: string | null;
  status: string;
  warnings: string[];
}

export async function listApplications(): Promise<ApplicationListResponse> {
  const res = await apiFetch(`${API_BASE}/applications`);
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(`API ${res.status}: ${text}`);
  }
  return res.json();
}

export async function getApplication(offerId: string): Promise<ApplicationItem> {
  const res = await apiFetch(`${API_BASE}/applications/${encodeURIComponent(offerId)}`);
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(`API ${res.status}: ${text}`);
  }
  return res.json();
}

export async function getApplicationHistory(offerId: string): Promise<ApplicationHistoryResponse> {
  const res = await apiFetch(
    `${API_BASE}/applications/${encodeURIComponent(offerId)}/history`
  );
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(`API ${res.status}: ${text}`);
  }
  return res.json();
}

export async function upsertApplication(
  payload: ApplicationCreatePayload
): Promise<ApplicationItem> {
  const res = await apiFetch(`${API_BASE}/applications`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(`API ${res.status}: ${text}`);
  }
  return res.json();
}

export async function patchApplication(
  offerId: string,
  payload: ApplicationUpdatePayload
): Promise<ApplicationItem> {
  const res = await apiFetch(
    `${API_BASE}/applications/${encodeURIComponent(offerId)}`,
    {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }
  );
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(`API ${res.status}: ${text}`);
  }
  return res.json();
}

export async function deleteApplication(offerId: string): Promise<void> {
  const res = await apiFetch(
    `${API_BASE}/applications/${encodeURIComponent(offerId)}`,
    { method: "DELETE" }
  );
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(`API ${res.status}: ${text}`);
  }
}

export async function prepareApplication(
  offerId: string,
  payload: PreparePayload = {}
): Promise<PrepareResponse> {
  const res = await apiFetch(
    `${API_BASE}/applications/${encodeURIComponent(offerId)}/prepare`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }
  );
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(`API ${res.status}: ${text}`);
  }
  return res.json();
}
